### PR TITLE
Properly fail promises on socket closure

### PR DESCRIPTION
**Pull request by ekhvalov**

### CODE REVIEW

File: src/BeanstalkClient.php [hunk 1]
```diff
@@ -49,6 +50,18 @@ public function __construct(string $uri) {
                 }
             }
         });
+        $this->connection->addEventHandler("close", function () {
+            if ($this->isConnected) {
```
Comment (kelunik) [inline src/BeanstalkClient.php]: Cloud you explain why we need this boolean?

Comment (ekhvalov) [reply src/BeanstalkClient.php]: This boolean shows state of connection. It needs to determine was a connection closed by client or by server.

Comment (ekhvalov) [reply src/BeanstalkClient.php]: You are right. It was unnecessary boolean.

File: test/BeanstalkClientConnectionLostTest.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        return [
+            "no timeout" => [null, 500, 600],
+            "one second timeout" => [1, 900, 1100],
+        ];
+    }
+
+    private function initServer(): Server {
+        for ($port = self::PORT_RANGE_MIN; $port <= self::PORT_RANGE_MAX; $port++) {
+            try {
+                return listen("tcp://127.0.0.1:". $port);
```
Comment (kelunik) [inline test/BeanstalkClientConnectionLostTest.php:64]: You can change it to `:0` so it'll automatically pick a free port.

File: test/BeanstalkClientConnectionLostTest.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        $beanstalk = new BeanstalkClient("tcp://". $this->server->getAddress());
+        $connectionClosePromise = call(function ($connectionCloseTimeout) {
+            yield new Delayed($connectionCloseTimeout);
+            $this->server->close();
+        }, $connectionCloseTimeout);
+
+        yield timeout(all([
+            $beanstalk->reserve($reserveTimeout),
+            $connectionClosePromise
+        ]), $testFailTimeout);
```
Comment (kelunik) [inline test/BeanstalkClientConnectionLostTest.php:51]: You can use [`setTimeout`](https://github.com/amphp/phpunit-util/blob/99db640d44395cf85289e1bfbde9fbed9f2a45b9/src/AsyncTestCase.php#L170) instead.

File: src/ConnectionLostException.php [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+namespace Amp\Beanstalk;
+
+class ConnectionLostException extends BeanstalkException {
```
Comment (kelunik) [inline src/ConnectionLostException.php:5]: I think `ConnectionClosedException` would be a better name.

**Comment by kelunik** [comment]
Thanks!

**Pull request merged**